syft/workers/plan.py

```diff
@@ -0,0 +1,261 @@
+        executed lazily
+        :param bin_message: the message of a command received
+        :return: the None message serialized to specify the command was received
+        """
+        (some_type, (msg_type, contents)) = sy.serde.deserialize(bin_message, detail=False)
+
+        if msg_type != MSGTYPE.OBJ:
+            self.plan.append(bin_message)
+            self.readable_plan.append((some_type, (msg_type, contents)))
+
+        # we can't receive the results of a plan without
+        # executing it. So, execute the plan.
+        if msg_type in (MSGTYPE.OBJ_REQ, MSGTYPE.IS_NONE, MSGTYPE.GET_SHAPE):
+            return self.__call__()
+
+        return sy.serde.serialize(None)
+
+    def build_plan(self, args):
+        """
+        The plan must be built with some input data, here args. When they
+        are provided, they are sent to the plan worker, which executes its
+        blueprint: each command of the blueprint is catched by _recv_msg
+        and is used to fill the plan
+        :param args: the input data
+        """
+        # The ids of args of the first call, which should be updated when
+        # the function is called with new args
+        self.arg_ids = list()
+
+        local_args = list()
+        for i, arg in enumerate(args):
+            self.owner.register_obj(arg)
+            arg = arg.send(self)
+            arg.child.garbage_collect_data = False
+            self.arg_ids.append(arg.id_at_location)
+            local_args.append(arg)
+
+        res_ptr = self.plan_blueprint(*local_args)
+        res_ptr.child.garbage_collect_data = False
+
+        # The id where the result should be stored
+        self.result_ids = [res_ptr.id_at_location]
+
+    def replace_ids(self, from_ids, to_ids):
+        """
+        Replace pairs of tensor ids in the plan stored
+        :param from_ids: the left part of the pair: ids to change
+        :param to_ids: the right part of the pair: ids to replace with
+        """
+        # for every pair of id
+        for i in range(len(from_ids)):
+            # for every message of the plan
+            for j, msg in enumerate(self.readable_plan):
+                # look for the old id and replace it with the new one
+                self.readable_plan[j] = replace_ids(
+                    obj=msg,
+                    change_id=from_ids[i],
+                    to_id=to_ids[i],
+                    from_worker=self.id,
+                    to_worker=self.owner.id,
+                )
+        return self
+
+    def replace_worker_ids(self, from_worker_id, to_worker_id):
+        """
+        Replace occurrences of from_worker_id by to_worker_id in the plan stored
+        """
+        self.readable_plan = replace_ids(
+            obj=self.readable_plan,
+            change_id=-1,
+            to_id=-1,
+            from_worker=from_worker_id,
+            to_worker=to_worker_id,
+        )
+
+    def __call__(self, *args, **kwargs):
```

Comment:
This is a good interface. I like that these are callables as it is very intuitive